% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/pir_rename_to_std.R
\name{pir_rename_to_std}
\alias{pir_rename_to_std}
\title{Rename the \code{pir_params} filenames to follow a standard naming scheme.}
\usage{
pir_rename_to_std(pir_params, folder_name)
}
\arguments{
\item{pir_params}{the parameters of \link[pirouette]{pirouette}.
They are created by \link{create_pir_params}.}

\item{folder_name}{name of the main folder}
}
\value{
a `pir_params`
}
\description{
By default, \code{pir_params} uses temporary filenames for all files.
For \link{pir_run}, when only a computer reads those filenames,
this is fine. This function conformizes the filenames to
a human-friendly form.
}
\details{
The standard naming scheme is this:

\itemize{
  \item \code{pir_params$alignment_params$fasta_filename}
    becomes \code{[folder_name]/alignment.fas}
  \item
    \code{pir_params$evidence_filename}
    becomes \code{[folder_name]/evidence.csv},
    if at least one evidence is measured
}

For the (zero or one) experiment at index \code{i}
that is generative:

\itemize{
  \item \code{pir_params$experiments[[i]]$beast2_options$input_filename}
    becomes \code{[folder_name]/gen.xml}
  \item
    \code{pir_params$experiments[[i]]$beast2_options$output_state_filename}
    becomes \code{[folder_name]/gen.xml.state}
  \item
    \code{pir_params$experiments[[i]]$beast2_options$input_filename}
    becomes \code{[folder_name]/gen_errors.csv},
  \item
    \code{pir_params$experiments[[i]]$inference_model$mcmc$tracelog$filename}
    becomes \code{[folder_name]/gen.log}
  \item
    \code{
      pir_params$experiments[[i]]$inference_model$mcmc$screenlog$filename
     }
    becomes \code{[folder_name]/gen.csv}
  \item
    \code{pir_params$experiments[[i]]$inference_model$mcmc$treelog$filename}
    becomes \code{[folder_name]/gen.trees}
  \item
    \code{pir_params$experiments[[i]]$est_evidence_mcmc$tracelog$filename}
    becomes \code{[folder_name]/gen_evidence.log}
  \item
    \code{pir_params$experiments[[i]]$est_evidence_mcmc$screenlog$filename}
    becomes \code{[folder_name]/gen_evidence.csv}
  \item
    \code{pir_params$experiments[[i]]$est_evidence_mcmc$treelog$filename}
    becomes \code{[folder_name]/gen_evidence.trees}
}

For the (zero or more) experiments at index \code{i}
that is a candidate:

\itemize{
  \item \code{pir_params$experiments[[i]]$beast2_options$input_filename}
    becomes \code{[folder_name]/best.xml}
  \item
    \code{pir_params$experiments[[i]]$beast2_options$output_state_filename}
    becomes \code{[folder_name]/best.xml.state}
  \item
    \code{pir_params$experiments[[i]]$beast2_options$input_filename}
    becomes \code{[folder_name]/best_errors.csv},
  \item
    \code{pir_params$experiments[[i]]$inference_model$mcmc$tracelog$filename}
    becomes \code{[folder_name]/best.log}
  \item
    \code{
      pir_params$experiments[[i]]$inference_model$mcmc$screenlog$filename
    }
    becomes \code{[folder_name]/best.csv}
  \item
    \code{pir_params$experiments[[i]]$inference_model$mcmc$treelog$filename}
    becomes \code{[folder_name]/best.trees}
  \item
    \code{pir_params$experiments[[i]]$est_evidence_mcmc$tracelog$filename}
    becomes \code{[folder_name]/best_evidence.log}
  \item
    \code{pir_params$experiments[[i]]$est_evidence_mcmc$screenlog$filename}
    becomes \code{[folder_name]/best_evidence.csv}
  \item
    \code{pir_params$experiments[[i]]$est_evidence_mcmc$treelog$filename}
    becomes \code{[folder_name]/best_evidence.trees}
}

If twinning is used:

\itemize{
  \item \code{pir_params$twinning_params$twin_tree_filename}
    becomes \code{[folder_name]/twin.newick}
  \item
    \code{pir_params$twinning_params$twin_alignment_filename}
    becomes \code{[folder_name]/alignment_twin.fas}
  \item
    \code{pir_params$twinning_params$twin_evidence_filename}
    becomes \code{[folder_name]/evidence_twin.csv},
    if at least one evidence is measured
}
}
\seealso{
Use \link{get_pir_params_filenames} to obtain all the filenames
}
