---
title: "article"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{article}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# BioRxiv article

```{r}
library(pirouette)
```

We show the usage of \verb;pirouette; by gradually introducing
its features.
First, to get an idea of the baseline error, 
we measure the Bayesian inference error when we start from a phylogeny 
generated under a known and standard tree model.
Second, to establish that the true tree prior is indeed the best, 
we also measure the inference error made by a
best candidate inference model.
Lastly, we quantify the impact of the tree prior in the Bayesian inference. 
We do so by separating the baseline error from the complete error 
running the pipeline starting from a tree generated by a non-standard tree model.

All the figures shown in this section are shown as-is, without any
aesthetical modifications. Figures showing the full workflow 
and tables showing the effective sample sized (a measure
of inference quality) can be found in the supplementary materials.

## The generative and inference models are equal and standard.

\verb;pirouette; quantifies the influence of a new tree prior on BEAST2's 
inference by measuring the discrepancy between a given/true tree and a posterior 
distribution of phylogenies obtained as result of the inference process. 
Due to stochasticity, posterior trees will generally differ from the given 
phylogeny $\tau_{\mathit{G}}$, even when the tree 
prior and alignment model used for inference are the same as those 
used to generate the alignment.
Measuring this difference allows us to know the baseline error
of the \verb;pirouette; pipeline. We therefore define as 'standard tree priors' 
all the tree priors that are available 
to be used within an inference model (see Table~\ref{tab:options}).

Now we can formulate the first example research question that \verb;pirouette; 
can answer: "What is the inference error made on phylogenies created by a 
standard diversification model?"

In this example we use a standard generative tree model $\mathit{p_{G}^0}$, 
namely the Yule (pure-birth) tree model. 
We choose to use a small tree with six taxa, to keep
the calculations short and the figure more readable.
We pick a crown age of ten time units. This value is 
completely arbitrary, but it ties in with the mutation rate 
used in simulating an alignment in the next step.

```{r create_yule_tree}
phylogeny <- create_yule_tree(n_taxa = 6, crown_age = 10)
```

The first step in \verb;pirouette; is to simulate a DNA alignment from the 
given phylogeny, as described in Subsection~\ref{subsec:pipeline}.
In this example, the root sequence consists of four blocks of 250 
mononucleotides each, while the per-nucleotide mutation rate is 
0.1 mutations per unit time.
We use a Jukes-Cantor (JC, \cite{jukes1969evolution}) nucleotide substitution model
and a strict clock model as these are the simplest.
A JC model assumes that mutation rates between nucleotides are equal and 
constant. 
A strict clock model assumes that the mutation rates 
of all lineages are equal and constant.

```{r create_alignment_params}
alignment_params <- create_alignment_params(
  sim_tral_fun = get_sim_tral_with_std_nsm_fun(
    mutation_rate = 0.1
  ),
  root_sequence = create_blocked_dna(length = 1000)
)
```

As the site and clock models used here are also the defaults, 
the function arguments can be safely omitted: we just explicitly show 
them for the sake of clarity.

In the second step we state our experiment.
We define an experiment $\mathit{X}$ as a combination of an inference model 
$\mathit{I}$ and conditions $\mathit{C}$.
In this example we choose $\mathit{I}$ to be the same inference model as 
the generative one,
i.e. the Yule tree prior $\mathit{p_{G}^0}$ and site 
and clock models defined in $\mathit{A}$, 
respectively Jukes-Cantor and strict clock.
We specify in $\mathit{C}$ that the experiment will always be run.

Listing~\ref{lst:create_gen_experiment_explicit} shows how to
set up this experiment:

```{r create_gen_experiment_explicit}
generative_experiment <- create_experiment(
  inference_conditions = create_inference_conditions(
    model_type = "generative", 
    run_if = "always"
  ), 
  inference_model = create_inference_model(
    tree_prior = create_yule_tree_prior(),
    clock_model = create_strict_clock_model(), 
    site_model = create_jc69_site_model(),
    mcmc = create_mcmc()
  )
)
```

Experiments must be bundled in a list to work, even if only one is provided, as 
in this case:

```{r create_gen_experiment}
experiments <- list(generative_experiment)
```

We also need to specify the error measurement parameters $\mathit{E}$.
Here we choose the default $\mathit{E}$, which has a burn-in fraction 
of 10\% and uses the nLTT statistic to
measure the difference between phylogenies. For clarity,
we create this setup explicitly here:

```{r create_error_measure_params}
error_measure_params <- create_error_measure_params(
  error_fun = get_nltt_error_fun(),
  burn_in_fraction = 0.1
)
```

We now have all the needed \verb;pirouette; parameters: the alignment 
parameters, the experiments and the error measure parameters.
These objects need to be bundled in a larger parameter structure, 
using \verb;create_pir_params;:

```{r create_pir_params}
pir_params <- create_pir_params(
  alignment_params = alignment_params,
  experiments = experiments,
  error_measure_params = error_measure_params
)
```

We can finally use the given Yule tree and \verb;pir_params; to measure the 
inference error made on phylogenies
created by a standard diversification model:

```{r pir_run}
errors <- pir_run(
  phylogeny = phylogeny,
  pir_params = pir_params
)
```

The error distribution can be plotted directly using \verb;pir_plot;:

```{r pir_plot}
pir_plot(errors)
```

The resulting error distribution, as shown Figure~\ref{fig:example_1},
shows the inference error 
when $\mathit{I}$ matches with the generative model given 
by $\mathit{A}$ and $\mathit{p_{G}^0}$.
This error distribution can serve as a control,
as it is obtained from a tree of which the generating tree model is standard and known.
Obtaining this control is in fact an inherent part of \verb;pirouette;. 
We call it 'twinning' and we will demonstrate it in section
\ref{Comparing to background noise} below.

## The inference model may differ from the generative model

In the previous example we selected the inference model $\mathit{I}$ to match 
with a known generative tree model $\mathit{p_{G}^0}$.
However, a novel tree prior $\mathit{p_{G}}$ is by (our) definition
non-standard, and hence not part of a standard inference 
model (i.e. an inference model using a standard tree prior).
In such a case, the question is which standard tree prior should be used in 
the inference. In this example, we do not only re-use our hand-picked 
inference model, but we also pick the best inference model from a set of 
inference models, i.e.  the model with 
the highest evidence measured by its marginal likelihood. 
We show the procedure using \verb;pirouette; to answer a second 
research question: "What is the inference error made on a novel phylogeny 
when using the best inference model, in comparison to a hand-picked model?

We will use the same tree as generated in \ref{lst:create_yule_tree}, as well 
as the same alignment parameters as shown 
in Listing~\ref{lst:create_alignment_params}.

Here we specify a different set of experiments: we
need to state that we already have an experiment for
the generative model, as well as that we want all the other inference models 
to compete. We call the competing models 'candidate models'.
As model selection is commonly performed on the full list of available 
candidate models, \verb;pirouette; has a dedicated function 
for this choice: \verb;create_all_experiments; creates a full set 
of 40 experiments, 
containing the inference models of all combinations of 4 nucleotide substitution model, 
2 clock models and 5 tree priors. All we need to add is to exclude the 
inference model in the generative experiment:

```{r create_all_experiments}
candidate_experiments <- create_all_experiments(
  exclude_model = generative_experiment$inference_model
)
```

We combine the generative and all the candidate models into one set of 
experiments:

```{r combine_all_experiments}
experiments <- c(
  list(generative_experiment),
  candidate_experiments
)
```

We can now create the complete \verb;pirouette; parameter set in the usual 
way (which is the same as Listing~\ref{lst:create_pir_params}, but using the 
defaults):

```{r create_pir_params_defaults}
pir_params <- create_pir_params(
  alignment_params = alignment_params,
  experiments = experiments
)
```

We run \verb;pirouette; (Listing~\ref{lst:pir_run}) 
and plot the results (Listing~\ref{lst:pir_plot}) in Figure~\ref{fig:example_2}.

```{r figure_2}
errors <- pir_run(
  phylogeny = phylogeny,
  pir_params = pir_params
)
pir_plot(errors)
```

## The generative model is non-standard}\label{Comparing to background noise

So far we have measured the inference error on a tree
generated according to a known (and standard) tree diversification model. 
The goal of \verb;pirouette; is to measure the expected impact of a novel tree prior.
To do so, in this example, we will use a tree generated by a non-standard 
tree diversification model, assumed to be closely related to the Yule tree prior.
We will use the diversity-dependent tree model \cite{DDD, etienne2011diversity},
which is a birth-death model with a speciation that is dependent on the
number of species.
We measure both the baseline 
error (that serves as a control) and the full inference error, because their 
difference shows the impact of the tree model.
We obtain the baseline error by using a twin tree (see Subsection~\ref{subsec:twinning}) for 
both the generative and best candidate models.
The research question this example answers is:
"What is the inference error made from a phylogeny, 
for both a generative model and the best candidate model?"

We start from the tree generated by a diversity-dependent model,
having six taxa and a crown age of ten:

```{r unknown_phylogeny}
# TODO: Use create_dd_tree
phylogeny  <- ape::read.tree(
  text = "(((A:8, B:8):1, C:9):1, ((D:8, E:8):1, F:9):1);"
)
```

Most of the other settings are the same as before: 
we reuse the alignment parameters (Listing~\ref{lst:create_alignment_params}), 
as well as the experiments (Listing~\ref{lst:create_all_experiments}).
This time, however, we enable twinning (see Subsection~\ref{subsec:twinning}),
by creating twinning parameters.
Creating these parameters is trivial with the default settings.
For clarity, however, we explicitly show the most important arguments:

```{r create_twinning_params}
twinning_params <- create_twinning_params(
  sim_twin_tree_fun = get_sim_bd_twin_tree_fun(),
  sim_twal_fun = get_sim_twal_with_std_nsm_fun()
)
```

We combine all the parameters using \verb;create_pir_params;:

```{r create_pir_params_with_twinning}
pir_params <- create_pir_params(
  alignment_params = alignment_params,
  experiments = experiments,
  twinning_params = twinning_params
)
```

We run \verb;pirouette; (Listing~\ref{lst:pir_run}) 
and plot the results (Listing~\ref{lst:pir_plot}).
The output is shown in Figure~\ref{fig:example_3}. 
While the error distributions using the best or generative model 
as inference model are very similar, 
the error distributions of the true tree are substantially 
larger than those of the twin tree. 
This is the error made by the mismatch
between the generating species tree model and the tree prior used in inference.

```{r figure_3}
errors <- pir_run(
  phylogeny = phylogeny,
  pir_params = pir_params
)
pir_plot(errors)
```
