---
title: "pirouette demo"
author: "Richel J.C. Bilderbeek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{pirouette demo}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Load library:

```{r}
library(pirouette)
```

Create the true phylogeny:

```{r}
sim_rng_seed <- 42
alignment_rng_seed <- 42
beast2_rng_seed <- 42
crown_age <- 15
use_simple_tree <- FALSE

set.seed(sim_rng_seed)
true_phylogeny <- NA
if (use_simple_tree) {
  true_phylogeny <- TreeSim::sim.bd.age(
    age = crown_age, 
    numbsim = 1, 
    lambda = 0.1, 
    mu = 0.0, 
    frac = 1.0, 
    mrca = TRUE
  )[[1]]
} else {
  sirg  <- 0.2
  siri  <- 0.2
  scr <- 0.01
  erg <- 0.0
  eri <- 0.0
  true_phylogeny <- PBD::pbd_sim(
    pars = c(sirg, scr, siri, erg, eri),
    age = crown_age,
    soc = 2 # Use crown age
  )$stree_youngest
}
```

```{r fig.width=7, fig.height=7}
ape::plot.phylo(true_phylogeny, main = "The 'true' phylogeny")
```

Now `pirouette` can create a Bayesian posterior from that same tree, 
in three ways:

 1. Let the crown age be estimated
 2. Let the posterior have a fixed crown age
 3. Let the Bayesian analysis have an MRCA prior on all taxa, that
      has a narrow distribution around the crown age

For this analysis, the same parameters will be used:

```{r}
n_nucleotides <- 200
mcmc <- create_mcmc(chain_length = 10000)

# per-nucleotide chance of a mutation per time unit
per_nuc_p_mut_per_tu <- 0.01
```

### 1. Let the crown age be estimated

```{r}
out <- out_1 <- pirouette::run(
  phylogeny = true_phylogeny,
  sequence_length = n_nucleotides,
  mcmc = mcmc,
  mutation_rate = per_nuc_p_mut_per_tu,
  alignment_rng_seed = alignment_rng_seed,
  beast2_rng_seed = beast2_rng_seed
)
```

The alignment it created can be visualized:

```{r fig.width=7, fig.height=7}
image(out$alignment)
```

The posterior trees `pirouette` created can be visualized:

```{r fig.width=7, fig.height=7}
trees <- out$trees
colors <- rep(c("red", "black"), each = length(out$trees))
for (i in seq(1, length(out$trees))) {
  trees <- c(trees, true_phylogeny)
}
babette::plot_densitree(
  trees, 
  width = 2, 
  consensus = true_phylogeny, 
  col = colors, 
  scaleX = TRUE,
  alpha = 10 / length(trees)
)
```

Comparing the nLTT plots:

```{r fig.width=7, fig.height=7}
nLTT::nltts_plot(phylogenies = out$trees, plot_nltts = TRUE, 
  main = "nLTTs of posterior trees (grey) and true tree (red)"
)
nLTT::nltt_lines(phy = true_phylogeny, col = "red",  lwd = 3)
```

### 2. Let the posterior have a fixed crown age

```{r}
out <- out_2 <- pirouette::run(
  phylogeny = true_phylogeny,
  sequence_length = n_nucleotides,
  mcmc = mcmc,
  mutation_rate = per_nuc_p_mut_per_tu,
  crown_age = crown_age,
  alignment_rng_seed = alignment_rng_seed,
  beast2_rng_seed = beast2_rng_seed
)
```

The alignment it created can be visualized:

```{r fig.width=7, fig.height=7}
image(out$alignment)
```

The posterior trees `pirouette` created can be visualized:

```{r fig.width=7, fig.height=7}
trees <- out$trees
colors <- rep(c("green", "black"), each = length(out$trees))
for (i in seq(1, length(out$trees))) {
  trees <- c(trees, true_phylogeny)
}
babette::plot_densitree(
  trees, 
  width = 2, 
  consensus = true_phylogeny, 
  col = colors,
  alpha = 10 / length(trees)
)
```

Comparing the nLTT plots:

```{r fig.width=7, fig.height=7}
nLTT::nltts_plot(phylogenies = out$trees, plot_nltts = TRUE, 
  main = "nLTTs of posterior trees (grey) and true tree (red)"
)
nLTT::nltt_lines(phy = true_phylogeny, col = "red",  lwd = 3)
```

### 3. Let the Bayesian analysis have an MRCA prior on all taxa

```{r}
out <- out_3 <- pirouette::run(
  phylogeny = true_phylogeny,
  sequence_length = n_nucleotides,
  mcmc = mcmc,
  mutation_rate = per_nuc_p_mut_per_tu,
  mrca_distr = beautier::create_normal_distr(
    mean = beautier::create_mean_param(value = crown_age),
    sigma = beautier::create_sigma_param(value = crown_age / 1000.0)
  ),
  alignment_rng_seed = alignment_rng_seed,
  beast2_rng_seed = beast2_rng_seed
)
```

The alignment it created can be visualized:

```{r fig.width=7, fig.height=7}
image(out$alignment)
```

The posterior trees `pirouette` created can be visualized:

```{r fig.width=7, fig.height=7}
trees <- out$trees
colors <- rep(c("blue", "black"), each = length(out$trees))
for (i in seq(1, length(out$trees))) {
  trees <- c(trees, true_phylogeny)
}
babette::plot_densitree(
  trees, 
  width = 2, 
  consensus = true_phylogeny, 
  col = colors,
  alpha = 10 / length(trees)
)
```

Comparing the nLTT plots:

```{r fig.width=7, fig.height=7}
nLTT::nltts_plot(phylogenies = out$trees, plot_nltts = TRUE, 
  main = "nLTTs of posterior trees (grey) and true tree (red)"
)
nLTT::nltt_lines(phy = true_phylogeny, col = "red",  lwd = 3)
```

## Comparing all trees

Prepare all trees and colors:

 * Red: trees with an estimated crown age (unscaled)
 * Green: trees with fixed crown age
 * Blue: trees with MRCA prior, with a distribution around the crown age
 * Black: true tree

```{r fig.width=7, fig.height=7}
trees <- c(out_1$trees, out_2$trees, out_3$trees)
colors <- rep(c("red", "green", "blue"), each = length(trees) / 3)
for (i in seq(1,50)) {
  trees <- c(trees, true_phylogeny)
  colors <- c(colors, "black")
}
```



```{r fig.width=7, fig.height=7}
phangorn::densiTree(trees, consensus = true_phylogeny, col = colors, alpha = 1)
phangorn::densiTree(trees, consensus = true_phylogeny, col = colors, alpha = 10 / length(trees))

phangorn::densiTree(trees, consensus = true_phylogeny, col = colors, scaleX = TRUE, alpha = 1)
phangorn::densiTree(trees, consensus = true_phylogeny, col = colors, scaleX = TRUE, alpha = 10 / length(trees))

```
