---
title: "pirouette demo"
author: "Richel J.C. Bilderbeek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{pirouette demo}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

The goal of `pirouette` is to estimate the error BEAST2 makes from a known 
phylogeny. This phylogeny can be created using a non-BEAST speciation model,
for example the PBD or MBD models.

```{r}
library(pirouette)
```

```{r}
if (!mauricer::is_beast2_pkg_installed("NS")) {
  mauricer::install_beast2_pkg("NS")
}
```

In this demo, we use a simple 'true' phylogeny:

```{r}
true_phylogeny <- ape::read.tree(text = "(((A:1, B:1):1, C:2):1, D:3);")
ape::plot.phylo(true_phylogeny, main = "The 'true' phylogeny")
```

As this vignette is a demo (and not a thorough research), 
we'll use a short alignment:

```{r}
alignment_params <- create_alignment_params(
  root_sequence = create_blocked_dna(length = 20),
  mutation_rate = 1.0 / beautier:::get_phylo_crown_age(true_phylogeny)
)
```

Also the inference is kept short, using a short MCMC chain:

```{r}
<<<<<<< HEAD
knitr::kable(head(errors))
```

Here we plot the error:

```{r}
pir_plot(errors)
```


```{r}
sim_rng_seed <- 42
alignment_rng_seed <- 42
beast2_rng_seed <- 42
crown_age <- 15
```

```{r fig.width=7, fig.height=7}
```

Now `pirouette` can create a Bayesian posterior from that same tree, 
in three ways:

 1. Let the crown age be estimated
 2. Let the posterior have a fixed crown age
 3. Let the Bayesian analysis have an MRCA prior on all taxa, that
      has a narrow distribution around the crown age

For this analysis, the same parameters will be used:

```
n_nucleotides <- 200
mcmc <- beautier::create_mcmc(chain_length = 10000)

# per-nucleotide chance of a mutation per time unit
per_nuc_p_mut_per_tu <- 0.01
```

### 1. Let the crown age be estimated

```
out <- out_1 <- pirouette::pir_run(
  phylogeny = true_phylogeny,
  sequence_length = NULL,
  root_sequence = create_blocked_dna(length = n_nucleotides),
  mcmc = mcmc,
  mutation_rate = per_nuc_p_mut_per_tu,
  alignment_rng_seed = alignment_rng_seed,
  beast2_rng_seed = beast2_rng_seed
=======
inference_param <- create_inference_param(
    mcmc = beautier::create_mcmc(chain_length = 2000, store_every = 1000)
>>>>>>> 7a192614618c42bf487743de7dfd2bd45b18c570
)
```

## Only use the generative model in inference

There are multiple ways to select for an inference model.
In the simplest case, which is the default, 
we will do our inference with the generative model only:

```{r}
model_select_params <- list(
  create_gen_model_select_param(
    alignment_params = alignment_params
  )
)
```

Here we measure the baseline error: the error that BEAST2 makes
when the inference model is the correct generative 
model (JC69 and strict clock):

```{r}
errors <- pir_run(
  true_phylogeny,
  alignment_params = alignment_params,
  model_select_params = model_select_params,
  inference_param = inference_param
)
```

Here we show the errors as a table:

```{r}
knitr::kable(head(errors))
```

Here we plot the error:

```{r fig.width=7, fig.height=7}
pir_plot(errors)
```

## Do inference with the generative and model with most evidence

Again, there are multiple ways to select for which inference model(s)
to use. Next to the known generative model (JC69 and strict clock),
here we use the inference model that has the highest evidence (also
called 'marginal likelihood'). 

Here we measure the baseline error, from the generative model,
and the error that BEAST2 makes
when the inference model is the one with most evidence:

```{r}
model_select_params = list(
  create_gen_model_select_param(
    alignment_params = alignment_params
  ),
  create_best_model_select_param(
    site_models = beautier::create_site_models()[3],
    clock_models = beautier::create_clock_models()[2],
    tree_priors = beautier::create_tree_priors()[5]
  )
)
```

Note that we select the model with the heighest evidence from a set of
just one model (JC69, strict, Yule).

Run this setup:

```{r}
errors <- pir_run(
  true_phylogeny,
  alignment_params = alignment_params,
  model_select_params = model_select_params,
  inference_param = inference_param
)
```

Here we show the errors as a table:

```{r}
knitr::kable(head(errors))
```

Here we plot the error:

```{r fig.width=7, fig.height=7}
pir_plot(errors)
```
