---
title: "Twinning"
author: "Richel J.C. Bilderbeek"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Twinning}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

`pirouette` has the option to investigate the error BEAST2 makes
on a twin tree. A twin tree is a tree with the same topology as the
original phylogeny, yet where the branch lengths follow a birth-death
branch length distribution.

```{r}
library(pirouette)
```

Twinning is useful to separate the effect of an unknown tree 
prior (i.e. speciation model) on a phylogeny's shape from the 
noise (the minimal error) made by BEAST2.

In this example, the following tree is used:

```{r}
phylogeny <- ape::read.tree(text = "((A:1, B:1):1, (C:1, D:1):1);")
ape::plot.phylo(phylogeny, main = "The True Phylogeny")
```

This phylogeny follows a speciation model unknown to BEAST2, as two
speciation events happened at exactly the same time. Such a phylogeny
will have a likelihood of zero for all of the BEAST2 tree priors.

Twinning:

```{r}
twinning_params <- pirouette::create_twinning_params(
  rng_seed = 42
)
```

```{r}
alignment_params <- pirouette::create_alignment_params(
  root_sequence = pirouette::create_blocked_dna(length = 100),
  mutation_rate = create_standard_mutation_rate,
  rng_seed = 314
)
```

We select our inference models in two ways:

  1. Using the generative model: we use the same site and clock model
     the alignment was created with. We do have to specify a BEAST2
     tree prior that is closest to what we think generated the phylogeny
  2. (for non-Windows users only) Using the inference model that
     has the most evidence (also: marginal likelihood) from a set of inference
     models. In this example, we pick all combinations of the one
     correct site model, the one correct clock model and two tree
     priors.

type       | run_if         | measure evidence | inference model
-----------|----------------|------------------|----------------
generative | always         |`TRUE`            |Birth-Death       
candidate  | best_candidate |`TRUE`            |Yule

```{r}
experiment_1 <- create_experiment(
  inference_conditions = create_inference_conditions(
    model_type = "generative",
    run_if = "always",
    do_measure_evidence = FALSE # Set to TRUE if UNIX 
  ), 
  inference_model = create_inference_model(
    tree_prior = create_bd_tree_prior(),
    mcmc = create_mcmc(chain_length = 3000, store_every = 1000)
  ),
  est_evidence_mcmc = create_nested_sampling_mcmc(epsilon = 100.0),
  beast2_options = create_beast2_options(rng_seed = 314)
)
if (rappdirs::app_dir()$os != "win") {
  experiment_2 <- create_experiment(
    inference_conditions = create_inference_conditions(
      model_type = "candidate",
      run_if = "best_candidate",
      do_measure_evidence = TRUE
    ),
    inference_model = create_inference_model(
      tree_prior = create_yule_tree_prior(),
      mcmc = create_mcmc(chain_length = 3000, store_every = 1000)
    ),
    est_evidence_mcmc = create_nested_sampling_mcmc(epsilon = 100.0),
    beast2_options = create_beast2_options(rng_seed = 314)
  )
  experiment_1$inference_conditions$do_measure_evidence <- TRUE
  experiments <- list(experiment_1, experiment_2)
} else {
  experiments <- list(experiment_1)  
}

pir_params <- create_pir_params(
  alignment_params = alignment_params,
  experiments = experiments,
  twinning_params = twinning_params
)
```

Run:

```{r}
df <- NULL
if (rappdirs::app_dir()$os != "win" && is_beast2_installed()) {
  df <- pirouette::pir_run(
    phylogeny = phylogeny,
    pir_params = pir_params
  ) 
} else {
  df <- create_test_pir_run_output()
}
```

Show as a table:

```{r}
knitr::kable(df)
```

Show as a figure:

```{r fig.width=7, fig.height=7}
pirouette::pir_plot(df)
```

## Appendix

```{r}
library(ggplot2)
```

### True tree

See true tree again:

```{r}
ape::plot.phylo(phylogeny, main = "The True Phylogeny")
```

See the alignment generated from the true tree:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  testit::assert(file.exists(pir_params$alignment_params$fasta_filename))
  ape::image.DNAbin(ape::read.FASTA(file = pir_params$alignment_params$fasta_filename))
}
```

See the posterior trees generated from the true alignment, for the generative model:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  testit::assert(file.exists(pir_params$experiments[[1]]$beast2_options$output_trees_filenames))
  babette::plot_densitree(
    tracerer::parse_beast_trees(pir_params$experiments[[1]]$beast2_options$output_trees_filenames)
  )
}
```

See the evidence of the true alignments for the models:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  knitr::kable(utils::read.csv(pir_params$evidence_filename))
}
```

See the posterior trees generated from the true alignment, for the model with the most evidence:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  trees_filename <- pir_params$experiments[[2]]$beast2_options$output_trees_filenames
  testit::assert(file.exists(trees_filename))
  babette::plot_densitree(tracerer::parse_beast_trees(trees_filename))
}
```

See the posterior parameter estimates generated from the true alignment, for the generative model:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  testit::assert(file.exists(pir_params$experiments[[1]]$beast2_options$output_log_filename))
  df <- tracerer::parse_beast_log(pir_params$experiments[[1]]$beast2_options$output_log_filename)
  ggplot(data = df, aes(x = Sample, y = likelihood)) + geom_line()
}
```

See the posterior parameter estimates generated from the true alignment, for the model with the most evidence:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  log_filename <- pir_params$experiments[[2]]$beast2_options$output_log_filename
  testit::assert(file.exists(log_filename))
  df <- tracerer::parse_beast_log(log_filename)
  ggplot(data = df, aes(x = Sample, y = likelihood)) + geom_line()
}
```

### Twin tree

See the twin tree:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  ape::plot.phylo(ape::read.tree(pir_params$twinning_params$twin_tree_filename), main = "The Twin Tree")
}
```

See the alignment generated from the twin tree:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  testit::assert(file.exists(pir_params$twinning_params$twin_alignment_filename))
  ape::image.DNAbin(ape::read.FASTA(file = pir_params$twinning_params$twin_alignment_filename))
}
```

See the posterior trees generated from the twin alignment, for the generative model:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  trees_filename <- to_twin_filename(pir_params$experiments[[1]]$beast2_options$output_trees_filenames)
  testit::assert(file.exists(trees_filename))
  babette::plot_densitree(
    tracerer::parse_beast_trees(trees_filename)
  )
}
```

See the evidence of the true alignments for the models:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  knitr::kable(utils::read.csv(pir_params$twinning_params$twin_evidence_filename))
}
```
See the posterior trees generated from the twin alignment, for the model with the most evidence:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  trees_filename <- to_twin_filename(pir_params$experiments[[2]]$beast2_options$output_trees_filenames)
  testit::assert(file.exists(trees_filename))
  babette::plot_densitree(tracerer::parse_beast_trees(trees_filename))
}
```

See the posterior parameter estimates generated from the true alignment, for the generative model:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  log_filename <- to_twin_filename(pir_params$experiments[[2]]$beast2_options$output_log_filename)
  testit::assert(file.exists(log_filename))
  df <- tracerer::parse_beast_log(log_filename)
  ggplot(data = df, aes(x = Sample, y = likelihood)) + geom_line()
}
```

See the posterior parameter estimates generated from the true alignment, for the model with the most evidence:

```{r}
if (rappdirs::app_dir()$os != "win" && beastier::is_beast2_installed()) {
  log_filename <- to_twin_filename(pir_params$experiments[[2]]$beast2_options$output_log_filename)
  testit::assert(file.exists(log_filename))
  df <- tracerer::parse_beast_log(log_filename)
  ggplot(data = df, aes(x = Sample, y = likelihood)) + geom_line()
}
```
